src/main/java/frc/robot/subsystems/VisionSubsystem.java

```diff
@@ -175,122 +113,114 @@ public VisionSubsystem(DrivebaseWrapper aprilTagsHelper) {
-    if (!MathUtil.isNear(Timer.getFPGATimestamp(), RawTimestampSeconds, 5.0)) {
-      return;
-    }
-    // updates estimated pose
-    var estimatedPose = estimator.update(result);
-    // if there is an actual pose then it gets a timestamp and pose3d
-    if (estimatedPose.isPresent()) {
-      var TimestampSeconds = estimatedPose.get().timestampSeconds;
-      var FieldPose3d = estimatedPose.get().estimatedPose;
-      // sets full pose3d for logging
-      rawFieldPose3dEntry.set(FieldPose3d);
-      // if you disabled vision then vision doesn't do anything anymore
-      if (disableVision.getBoolean(false)) {
+  private void processLimelight(
+      LLCamera camera, StructPublisher<Pose3d> rawFieldPoseEntry, RawFiducial rf) {
+    if (disableVision.getBoolean(false)) return;
+    PoseEstimate estimate = camera.getPoseEstimate();
+
+    if (estimate != null) {
+      if (estimate.tagCount == 0) {
         return;
       }
-      //////// below is if a tag is a specific tag id don't read this result
-      // if (BadAprilTagDetector(result)) {
-      //   return;
+
+      double rawTimestampSeconds = estimate.timestampSeconds;
+      Pose3d fieldPose3d = camera.getPose3d();
+      boolean pose_bad = false;
+      rawFieldPoseEntry.set(fieldPose3d);
+      // distance to closest fiducial
+      double distanceMeters = Distance;
+      if (estimate.tagCount > 0) {
+        Optional<Pose3d> tagPose = fieldLayout.getTagPose(rf.id);
+        if (tagPose.isPresent()) {
+          distanceMeters = rf.distToCamera;
+        }
+      }
+
+      if (!MathUtil.isNear(0, fieldPose3d.getZ(), 0.10)
+          || !MathUtil.isNear(0, fieldPose3d.getRotation().getX(), Units.degreesToRadians(8))
+          || !MathUtil.isNear(0, fieldPose3d.getRotation().getY(), Units.degreesToRadians(8))
+          || estimate.tagCount == 1 && rf.distToCamera > 0.3) {
+        pose_bad = true;
+      }
+
+      // // filter invalid tags by alliance reef
+      // if (estimate.avgTagID >= 0 && isBadAprilTagForAlliance(estimate.avgTagID)) {
+      // return;
       // }
-      // tolerances for rotation if its outside of tolerances then return null
-      if (!MathUtil.isNear(0, FieldPose3d.getZ(), 0.10)
-          || !MathUtil.isNear(0, FieldPose3d.getRotation().getX(), Units.degreesToRadians(8))
-          || !MathUtil.isNear(0, FieldPose3d.getRotation().getY(), Units.degreesToRadians(8))) {
-        return;
+      if (!pose_bad) {
+        double stdDevFactor = Math.pow(estimate.avgTagDist, 2.0) / estimate.tagCount;
+        double linearStdDev = 0.02 * stdDevFactor;
+        double angularStdDev = 0.06 * stdDevFactor;
+        aprilTagsHelper.addVisionMeasurement(
+            fieldPose3d.toPose2d(),
+            rawTimestampSeconds,
+
+            //// Use one of these, first one is current second is what advantage kit example is
+            // DISTANCE_SC_STANDARD_DEVS.times(Math.max(0, distanceMeters -
+            // 1)).plus(STANDARD_DEVS));
```

Comment:
Please do the math (Excel is good at this) to compare the two methods for various tag distances & see how that affects the values in the vector below.